### ISSUE TITLE
Model Overfitting

### ISSUE THREAD

**Issue by basque21** [amaiya/ktrain#118]
Hi, I'm trying to train a text classifier, and I'm getting a huge overfitting. Do you know what could be the reason? My model is:

```
trn, val, preproc = text.texts_from_array(X, y_train,
                                           class_names=[0,1,2],
                                           max_features=1000, maxlen=250, 
                                           ngram_range=3, 
                                           val_pct=0.1,
                                           lang='es',
                                           preprocess_mode='standard',
                                           )
```

```
task: text classification
language: es
Word Counts: 1434105
Nrows: 238255
238255 train sequences
train sequence lengths:
	mean : 929
	95percentile : 2181
	99percentile : 3628
Adding 3-gram features
max_features changed to 42437014 with addition of ngrams
Average train sequence length with ngrams: 2784
train (w/ngrams) sequence lengths:
	mean : 2785
	95percentile : 6540
	99percentile : 10882
x_train shape: (238255,250)
y_train shape: (238255, 3)
26473 test sequences
test sequence lengths:
	mean : 936
	95percentile : 2168
	99percentile : 3715
Average test sequence length with ngrams: 2689
test (w/ngrams) sequence lengths:
	mean : 2690
	95percentile : 6261
	99percentile : 10677
x_test shape: (26473,250)
y_test shape: (26473, 3)
```
And then when I do `learner.fit(0.001, 3, cycle_len=1, cycle_mult=2)`, I get:

![image](https://user-images.githubusercontent.com/26438571/79064623-c6cf9980-7caa-11ea-8bf9-a9a23b73cf46.png)

Thanks!

**Comment by marcturpin91** [comment]
One thing I've noticed in your data is that your mean sequence is 2785 words, which is way above the sequence length you are actually processing (maxlen=250). This essentially means you are, on average, learning based on less than 10% of the data in your samples. 

Since BERT models limit your sequence length (words it can process) to 512, you will be forced to choose something that will truncate your data, as you have currently done by choosing 250 max len. I would suggest you find a way to make your samples smaller, and make sure the data that is being learned on is meaningful. This might help the overfitting issue. 

Some methods that have been suggested to deal with this issue is to select first X amount of words from the beginning, and Y amount of words from the end of your samples. You can also use an algorithm to summarize your sample so it would cut down on the size of each sample. You can split your samples into chunks. If required, you could combine the scores that belong to the same document at the end in order to get a proper final label / classification.

**Comment by marcturpin91** [comment]
Other ideas to look at:
1. Balance the dataset

2. Check to make sure your trn, val have roughly the same proportion of classes

3. Use predictor.predict_proba(<some test sample>) to debug. You can test a few samples and see if the probabilities are actually changing with respect to the sample provided. If they're not, your model is probably just guessing the most common class found in it's validation set.

4. Test your model on a smaller subset of your dataset to save time and debug. Once you get something good, you can move on to your full dataset. Make sure that the subset you are using is representative of your full dataset though. If you need to balance, balance the subset.

5. Change the learning rate (you are currently using 0.001). Maybe try 0.0001 and 0.01 to see if you get different results. You can use the following code to get a visual estimate of the ideal learning rate:

```
learner.lr_find(max_epochs=2)
# learner.lr_plot()  # only use these two last lines if you are running in some IDE and not a notebook
# pyplot.show()
```

6. Change the batch size

**Comment by amaiya** [comment]
@marcturpin91 provided some great answers.  I try to add a few comments here.

I haven't seen your dataset or the model you're using.  It looks like you're using non-transformer model perhaps because training a transformer model like BERT on such a large training set is taking too long.  If this is the case,  you can try downsampling your training set and using DistilBert, as Transformer models tend to generalize better.  You may not need that many training samples as you are currently using, as redundancy may be present in your dataset. There are lots of different ways to downsample such as [this technique](https://lig-membres.imag.fr/portet/publications/07idamap.pdf) or using [active learning](https://en.wikipedia.org/wiki/Active_learning_(machine_learning)).    Downsampling may also help in simply creating a smaller dataset to test other methods to reduce overfitting mentioned below. You can always got back to the original size once you figure out what works.

Other techniques include the following and include:
1. Add regularization such as dropout or L2/L1/weight_decay.
2. reduce complexity of your model

For weight decay, you can use `learner.set_weight_decay(wd=0.01)` to set the [weight decay](https://towardsdatascience.com/this-thing-called-weight-decay-a7cd4bcfccab) to 0.01. (or some other value)  Or, you can increase the weight decay if necessary, if 0.01 is too low or high.  As of the latest version of *ktrain* this will swap out the vanilla Adam in Keras with the [AdamW optimizer](https://arxiv.org/abs/1711.05101). 

As for other regularization techniques that involve adjusting the layers of the model like adding dropout layers, changing size of hidden units, batchnorm, etc., it's not clear if you're using your own Keras model or one of the models included in *ktrain*.  Even if the latter, you can copy the model-building code form ktrain/text/models and add dropout layers or add L2 `kernel_regularizers` to layers of the model, etc. to layers, etc.  In theory, weight decay and adding L2 regularization to each layer are equivalent, but it may behave differently in practice due to different factors in Keras.

See also this [tutorial on overfitting/underfitting](https://keras.rstudio.com/articles/tutorial_overfit_underfit.html).  This tutorial uses Keras for R, but everything still applies to `tf.keras`.

**Comment by basque21** [comment]
Hi, thanks for your answers @marcturpin91  and @amaiya ! I'll try to answer to your points:

1. - I'm getting only 250 words because of time and memory reducing for trials. Additionally I'm almost sure that the important part of the text is at the beginning... Is ktrain truncating getting the first 250 words or the last 250s?

2. - I checked that my trn, val sets has same proportion of classes. 

3. - I tried with some other learning rates and I'm getting almost the same result.

4. - I have also tried with my own Keras models and added L2/L1 and Dropout, but with the same result.

5. - Now I'm trying some techniques to balance the dataset, but not great news with Random Under/Over Samping, and others are getting a lot of time.

I'm going to explain a bit my dataset and say what I guess is the problem: There are 3 classes with 10-30-60% distribution. The classes are about importance of the documents, 10% class is the most relevant, 30% the second and 60% the last. I think that there are a lot of documents in class 3 that are similar to others in class 1, but were assigned there because of some company policy. Maybe it would be better to try some one-class learning for class 1 and 2 and find some threshold? What do you think?

Thanks a lot!!

**Comment by amaiya** [comment]
Hi @basque21 It's hard to say what's going on without access to the dataset.  But, if classes are similar then the model might be fitting to the noise.  I'll close the issue for now.
